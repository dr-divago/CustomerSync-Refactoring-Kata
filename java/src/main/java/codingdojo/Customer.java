package codingdojo;

import org.immutables.value.Value;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Value.Immutable
public abstract class Customer {
    public abstract Optional<String> externalId();
    public abstract Optional<String> masterExternalId();
    public abstract Optional<Address> address();
    public abstract Optional<String> preferredStore();
    public abstract List<ShoppingList> shoppingLists();
    public abstract Optional<String> internalId();
    public abstract Optional<String> name();
    public abstract CustomerType customerType();
    public abstract Optional<String> companyNumber();


    public boolean isInternal() {
        return internalId().isPresent();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Customer)) return false;
        Customer customer = (Customer) o;
        return Objects.equals(externalId(), customer.externalId()) &&
                Objects.equals(masterExternalId(), customer.masterExternalId()) &&
                Objects.equals(companyNumber(), customer.companyNumber());
    }

    @Override
    public int hashCode() {
        return Objects.hash(externalId(), masterExternalId(), companyNumber());
    }
}
