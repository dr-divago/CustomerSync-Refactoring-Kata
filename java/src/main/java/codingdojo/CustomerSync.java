package codingdojo;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;

public class CustomerSync {

    private final CustomerDataAccess customerDataAccess;

    public CustomerSync(CustomerDataLayer customerDataLayer) {
        this(new CustomerDataAccess(customerDataLayer));
    }

    public CustomerSync(CustomerDataAccess db) {
        this.customerDataAccess = db;
    }

    public boolean syncWithDataLayer(ExternalCustomer externalCustomer) {

        CustomerMatches customerMatches = externalCustomer.isCompany() ? loadCompany(externalCustomer) : loadPerson(externalCustomer);

        Customer customer = customerMatches.customer()
          .orElse(ImmutableCustomer.builder()
            .externalId(externalCustomer.externalId())
            .masterExternalId(externalCustomer.externalId())
            .customerType(externalCustomer.isCompany() ? CustomerType.COMPANY : CustomerType.PERSON)
            .build());


        Customer newCustomer = populateFields(externalCustomer, customer);
        newCustomer = updateRelations(externalCustomer, newCustomer);
        Collection<Customer> duplicates = customerMatches.duplicates();
        if (!duplicates.isEmpty()) {
          for (Customer duplicate : customerMatches.duplicates()) {
            updateDuplicate(externalCustomer, duplicate);
          }
        }
        boolean created = false;
        if (!newCustomer.isInternal()) {
            newCustomer = createCustomer(newCustomer);
            created = true;
        } else {
            newCustomer = updateCustomer(newCustomer);
        }



        updateCustomer(newCustomer);

        return created;
    }

    private Customer updateRelations(ExternalCustomer externalCustomer, Customer customer) {
        List<ShoppingList> consumerShoppingLists = externalCustomer.shoppingLists();
        for (ShoppingList consumerShoppingList : consumerShoppingLists) {
          List<ShoppingList> newList = new ArrayList<>(customer.shoppingLists());
          newList.add(consumerShoppingList);
          Customer updateCustomer = ImmutableCustomer.copyOf(customer).withShoppingLists(newList);
          customer = this.customerDataAccess.updateShoppingList(updateCustomer, consumerShoppingList);
        }
        return customer;
    }

    private Customer updateCustomer(Customer customer) {
        return this.customerDataAccess.updateCustomerRecord(customer);
    }

    private Customer updateDuplicate(ExternalCustomer externalCustomer, Customer duplicate) {
      Customer newCustomer = ImmutableCustomer.copyOf(duplicate)
          .withName(externalCustomer.name());

        return newCustomer.isInternal() ? updateCustomer(newCustomer) : createCustomer(newCustomer);

    }

    private Customer createCustomer(Customer customer) {
        return this.customerDataAccess.createCustomerRecord(customer);
    }

    private Customer populateFields(ExternalCustomer externalCustomer, Customer customer) {
      return ImmutableCustomer.copyOf(customer)
        .withName(externalCustomer.name())
        .withCompanyNumber(externalCustomer.companyNumber())
        .withCustomerType(externalCustomer.isCompany() ? CustomerType.COMPANY : CustomerType.PERSON)
        .withAddress(externalCustomer.address())
        .withPreferredStore(externalCustomer.preferredStore());
    }

    public CustomerMatches loadCompany(ExternalCustomer externalCustomer) {
        return customerDataAccess.loadCompanyCustomer(externalCustomer.externalId(), externalCustomer.companyNumber().orElseThrow(IllegalStateException::new));
    }


  public CustomerMatches loadPerson(ExternalCustomer externalCustomer) {
        final String externalId = externalCustomer.externalId();

        CustomerMatches customerMatches = customerDataAccess.loadPersonCustomer(externalId);

        if (customerMatches.customer().isPresent()) {
            if (!CustomerType.PERSON.equals(customerMatches.customer().get().customerType())) {
                throw new ConflictException("Existing customer for externalCustomer " + externalId + " already exists and is not a person");
            }
        }

        return customerMatches;
    }


}
