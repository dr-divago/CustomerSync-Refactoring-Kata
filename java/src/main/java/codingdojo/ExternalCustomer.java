package codingdojo;


import org.immutables.value.Value;

import java.util.List;
import java.util.Optional;

@Value.Immutable
public abstract class ExternalCustomer {
    public abstract Address address();
    public abstract String name();
    public abstract Optional<String> preferredStore();
    public abstract List<ShoppingList> shoppingLists();
    public abstract String externalId();
    public abstract Optional<String> companyNumber();

  public boolean isCompany() {
    return companyNumber().isPresent();
  }
}
