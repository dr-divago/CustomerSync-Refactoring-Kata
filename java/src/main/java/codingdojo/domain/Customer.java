package codingdojo.domain;

import org.immutables.value.Value;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Value.Immutable
public abstract class Customer {
    public abstract Optional<String> externalId();
    public abstract Optional<String> masterExternalId();
    public abstract Optional<Address> address();
    public abstract Optional<String> preferredStore();
    public abstract List<ShoppingList> shoppingLists();
    public abstract Optional<String> internalId();
    public abstract Optional<String> name();
    public abstract CustomerType customerType();
    public abstract Optional<String> companyNumber();
    public abstract Optional<Integer> bonusPoints();


    public boolean isInternal() {
        return internalId().isPresent();
    }

    public boolean isNotInternal() {
        return !isInternal();
    }

    public boolean isPerson() {
        return customerType().equals(CustomerType.PERSON);
    }

    public boolean isNotPerson() {
        return !isPerson();
    }

    public boolean isCompany() {
        return customerType().equals(CustomerType.COMPANY);
    }

    public boolean isNotCompany() {
        return !isCompany();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Customer)) return false;
        Customer customer = (Customer) o;
        return Objects.equals(externalId(), customer.externalId()) &&
                Objects.equals(masterExternalId(), customer.masterExternalId()) &&
                Objects.equals(companyNumber(), customer.companyNumber());
    }

    @Override
    public int hashCode() {
        return Objects.hash(externalId(), masterExternalId(), companyNumber());
    }
}
