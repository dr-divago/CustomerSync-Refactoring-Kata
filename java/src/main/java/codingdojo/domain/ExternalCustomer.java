package codingdojo.domain;


import java.util.List;
import java.util.Optional;
import org.immutables.value.Value;

@Value.Immutable
public abstract class ExternalCustomer {
    public abstract Address address();
    public abstract String name();
    public abstract Optional<String> preferredStore();
    public abstract List<ShoppingList> shoppingLists();
    public abstract String externalId();
    public abstract Optional<String> companyNumber();
    public abstract Optional<Integer> bonusPoints();

  public boolean isCompany() {
    return companyNumber().isPresent();
  }

  public boolean isPerson() {return !isCompany(); }
}
