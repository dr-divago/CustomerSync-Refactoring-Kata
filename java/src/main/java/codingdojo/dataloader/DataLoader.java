package codingdojo.dataloader;

import codingdojo.dao.CustomerDataLayer;
import codingdojo.domain.CustomerMatches;
import codingdojo.domain.ExternalCustomer;

public interface DataLoader {
  CustomerMatches load(ExternalCustomer externalCustomer);


  static DataLoader from(final ExternalCustomer externalCustomer, final CustomerDataLayer customerDataLayer) {
    if (externalCustomer.isCompany()) {
      return new CompanyDataLoader(customerDataLayer);
    }
    if (externalCustomer.isPerson()) {
      return new PersonDataLoader(customerDataLayer);
    }
    throw new IllegalArgumentException("Unknown customer type");
  }
}
