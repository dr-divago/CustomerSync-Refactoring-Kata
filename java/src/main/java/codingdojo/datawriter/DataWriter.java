package codingdojo.datawriter;

import codingdojo.domain.Customer;
import java.util.Collection;

public interface DataWriter {
  enum ACTION {
    CREATE,
    UPDATE
  }
  ACTION write(Customer customer);
  void write(Collection<Customer> duplicates);
}
