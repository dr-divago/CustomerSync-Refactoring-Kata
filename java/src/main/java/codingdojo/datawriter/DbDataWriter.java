package codingdojo.datawriter;

import codingdojo.dao.CustomerDataLayer;
import codingdojo.domain.Customer;
import codingdojo.domain.ImmutableCustomer;
import codingdojo.domain.ShoppingList;
import java.util.Collection;
import java.util.List;

public class DbDataWriter implements DataWriter {

  private final CustomerDataLayer customerDataLayer;

  public DbDataWriter(CustomerDataLayer customerDataLayer) {
    this.customerDataLayer = customerDataLayer;
  }

  @Override
  public ACTION write(Customer customer) {
    return customer.isInternal()
      ? updateCustomer(customer)
      : createCustomerRecord(customer);

  }

  @Override
  public void write(Collection<Customer> duplicates) {
      duplicates.stream()
        .filter(Customer::isInternal)
        .forEach(this.customerDataLayer::updateCustomerRecord);

      duplicates.stream()
        .filter(Customer::isNotInternal)
        .forEach(this.customerDataLayer::createCustomerRecord);
  }

  private ACTION updateCustomer(Customer customer) {
    this.updateCustomerRecord(customer);
    this.updateShoppingList(customer, customer.shoppingLists());
    return ACTION.UPDATE;
  }

  private ACTION createCustomerRecord(Customer customer) {
    Customer newCustomer = ImmutableCustomer
      .copyOf(customer)
      .withInternalId("fake internalId");
    customerDataLayer.createCustomerRecord(newCustomer);
    return ACTION.CREATE;
  }

  public Customer updateCustomerRecord(final Customer customer) {
    return customerDataLayer.updateCustomerRecord(customer);
  }

  public Customer updateShoppingList(
    final Customer customer,
    final List<ShoppingList> consumerShoppingList) {

    consumerShoppingList.forEach(customerDataLayer::updateShoppingList);
    return customerDataLayer.updateCustomerRecord(customer);
  }

}
