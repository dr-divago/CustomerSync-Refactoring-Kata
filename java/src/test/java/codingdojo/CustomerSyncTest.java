package codingdojo;

import static org.junit.jupiter.api.Assertions.assertEquals;

import codingdojo.dataloader.DataLoader;
import codingdojo.datawriter.DataWriter;
import codingdojo.datawriter.DataWriter.ACTION;
import codingdojo.datawriter.DbDataWriter;
import codingdojo.domain.Customer;
import codingdojo.domain.CustomerType;
import codingdojo.domain.ExternalCustomer;
import codingdojo.domain.ImmutableAddress;
import codingdojo.domain.ImmutableCustomer;
import codingdojo.domain.ImmutableExternalCustomer;
import codingdojo.domain.ShoppingList;
import codingdojo.service.ConflictException;
import codingdojo.service.CustomerSync;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;


public class CustomerSyncTest {

    /**
     * The external record already exists in the customer db, so no need to create it.
     * There is new data in some fields, which is merged in.
     */

    private DataWriter dataWriter;
    private FakeDatabase db;

    @BeforeEach
    public void setUp() {
      db = new FakeDatabase();
      dataWriter = new DbDataWriter(db);
    }

    @Test
    public void syncCompanyByExternalId(){
        String externalId = "12345";
        String companyNumber = "470813-8895";

        ExternalCustomer externalCustomer = createExternalCompany(externalId, companyNumber);

        Customer customer = createCustomerWithSameCompanyAs(externalCustomer,
          Optional.of(externalId), Optional.of(companyNumber), Collections.emptyList(), Optional.empty());

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);
        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.UPDATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncPrivatePersonByExternalId(){
        String externalId = "12345";

        ExternalCustomer externalCustomer = createExternalPrivatePerson(externalId);

        Customer customer = ImmutableCustomer.builder()
                .externalId(externalId)
                .internalId("67576")
                .customerType(CustomerType.PERSON)
                .bonusPoints(0)
                .build();

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.UPDATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncShoppingLists(){
        String externalId = "12345";
        String companyNumber = "470813-8895";

        ExternalCustomer externalCustomer = createExternalCompany(externalId, companyNumber);

        ShoppingList shoppingList = new ShoppingList("eyeliner", "blusher");
        Customer customer = createCustomerWithSameCompanyAs(externalCustomer,
          Optional.of(externalId),
          Optional.of(companyNumber),
          Arrays.asList(shoppingList),
          Optional.empty());

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.UPDATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncNewCompanyCustomer(){
        String externalId = "12345";
        String companyNumber = "470813-8895";
        ExternalCustomer externalCustomer = createExternalCompany(externalId, companyNumber);

        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.CREATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncNewPrivateCustomer(){
        String externalId = "12345";
        ExternalCustomer externalCustomer = createExternalPrivatePerson(externalId);

        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.CREATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void conflictExceptionWhenExistingCustomerIsPerson() {
        String externalId = "12345";
        String companyNumber = "470813-8895";

        ExternalCustomer externalCustomer = createExternalCompany(externalId,companyNumber);

        Customer customer = ImmutableCustomer.builder()
                .externalId(externalId)
                .internalId("45435")
                .customerType(CustomerType.PERSON)
                .bonusPoints(0)
                .build();

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        Assertions.assertThrows(ConflictException.class, () -> {
            sut.syncWithDataLayer(externalCustomer);
        }, printAfterState(db, toAssert).toString());

        Approvals.verify(toAssert);
    }

    @Test
    public void syncByExternalIdButCompanyNumbersConflict(){
        String externalId = "12345";
        String companyNumber = "470813-8895";

        ExternalCustomer externalCustomer = createExternalCompany(externalId, companyNumber);

        Customer customer = createCustomerWithSameCompanyAs(externalCustomer,
          Optional.of(externalId),
          Optional.of("000-3234"),
          Collections.emptyList(), Optional.empty());

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.CREATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }


    @Test
    public void syncByCompanyNumber(){
        String companyNumber = "12345";

        ExternalCustomer externalCustomer = createExternalCompany("12345", companyNumber);

        Customer customer = createCustomerWithSameCompanyAs(externalCustomer,
          Optional.empty(),
          Optional.of(companyNumber),
          Arrays.asList(new ShoppingList("eyeliner", "mascara", "blue bombe eyeshadow")),
          Optional.empty());

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.UPDATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    @Test
    public void syncByCompanyNumberWithConflictingExternalId(){
        String companyNumber = "12345";
        String externalId = "45646";

        ExternalCustomer externalCustomer = createExternalCompany(externalId, companyNumber);

        Customer customer = createCustomerWithSameCompanyAs(externalCustomer,
          Optional.of("conflicting id"),
          Optional.of(companyNumber),
          Collections.emptyList(),
          Optional.empty());

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        Assertions.assertThrows(ConflictException.class, () -> {
            sut.syncWithDataLayer(externalCustomer);
        }, printAfterState(db, toAssert).toString());

        Approvals.verify(toAssert);
    }

    @Test
    public void conflictExceptionWhenExistingCustomerIsCompany() {
        String externalId = "12345";

        ExternalCustomer externalCustomer = createExternalPrivatePerson(externalId);

        Customer customer = ImmutableCustomer.builder()
          .externalId(externalId)
          .internalId("45435")
          .companyNumber("32423-342")
          .customerType(CustomerType.COMPANY)
          .bonusPoints(0)
          .build();

        db.addCustomer(customer);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        Assertions.assertThrows(ConflictException.class, () -> {
            sut.syncWithDataLayer(externalCustomer);
        }, printAfterState(db, toAssert).toString());

        Approvals.verify(toAssert);
    }

    @Test
    public void syncCompanyByExternalIdWithNonMatchingMasterId(){
        String externalId = "12345";
        String companyNumber = "470813-8895";

        ExternalCustomer externalCustomer = createExternalCompany(externalId, companyNumber);

        Customer customer = createCustomerWithSameCompanyAs(externalCustomer,
          Optional.of(externalId),
          Optional.of(companyNumber),
          Collections.emptyList(),
          Optional.of("company 1"));

        Customer customer2 = ImmutableCustomer.builder()
                .companyNumber(externalCustomer.companyNumber())
                .internalId("45435234")
                .customerType(CustomerType.COMPANY)
                .masterExternalId(externalId)
                .name("company 2")
                .build();

        db.addCustomer(customer);
        db.addCustomer(customer2);
        DataLoader dataLoader = DataLoader.from(externalCustomer, db);
        CustomerSync sut = new CustomerSync(dataLoader, dataWriter);

        StringBuilder toAssert = printBeforeState(externalCustomer, db);

        // ACT
        ACTION created = sut.syncWithDataLayer(externalCustomer);

        assertEquals(created, ACTION.UPDATE);
        printAfterState(db, toAssert);
        Approvals.verify(toAssert);
    }

    private ExternalCustomer createExternalPrivatePerson(String externalId) {
        return ImmutableExternalCustomer.builder()
                .externalId(externalId)
                .name("Joe Bloggs")
                .address(ImmutableAddress.builder().street("123 main st").city("Stockholm").postalCode("SE-123 45").build())
                .preferredStore("Nordstan")
                .addShoppingLists(new ShoppingList("lipstick", "foundation"))
                .bonusPoints(50)
                .build();
    }


    private ExternalCustomer createExternalCompany(String externalId, String companyNumber) {
        return ImmutableExternalCustomer.builder()
                .externalId(externalId)
                .name("Acme Inc.")
                .address(ImmutableAddress.builder().street("123 main st").city("Helsingborg").postalCode("SE-123 45").build())
                .companyNumber(companyNumber)
                .preferredStore(Optional.empty())
                .bonusPoints(0)
                .addShoppingLists(new ShoppingList("lipstick", "blusher"))
                .build();
    }

    private Customer createCustomerWithSameCompanyAs(ExternalCustomer externalCustomer,
                                                     Optional<String> externalId,
                                                     Optional<String> companyNumber,
                                                     List<ShoppingList> shoppingList,
                                                     Optional<String> name) {
      return ImmutableCustomer.builder()
        .externalId(externalId)
        .companyNumber(externalCustomer.companyNumber().orElse(""))
        .internalId("45435")
        .customerType(CustomerType.COMPANY)
        .companyNumber(companyNumber.orElse(""))
        .shoppingLists(shoppingList)
        .preferredStore(Optional.empty())
        .bonusPoints(0)
        .name(name)
        .build();
    }

    private StringBuilder printBeforeState(ExternalCustomer externalCustomer, FakeDatabase db) {
        StringBuilder toAssert = new StringBuilder();
        toAssert.append("BEFORE:\n");
        toAssert.append(db.printContents());

        toAssert.append("\nSYNCING THIS:\n");
        toAssert.append(ExternalCustomerPrinter.print(externalCustomer, ""));
        return toAssert;
    }

    private StringBuilder printAfterState(FakeDatabase db, StringBuilder toAssert) {
        toAssert.append("\nAFTER:\n");
        toAssert.append(db.printContents());
        return toAssert;
    }
}
