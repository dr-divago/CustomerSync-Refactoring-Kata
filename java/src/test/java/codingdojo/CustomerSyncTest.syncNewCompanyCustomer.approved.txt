BEFORE:
Fake Database.
All Customers {

}
All Shopping Lists
[]
SYNCING THIS:
ExternalCustomer {
    externalId='12345'
    companyNumber='Optional[470813-8895]'
    name='Acme Inc.'
    preferredStore='Optional.empty'
    bonusPoints='Optional[0]'
    address='123 main st, SE-123 45 Helsingborg'
    shoppingLists=[[lipstick, blusher]]
}
AFTER:
Fake Database.
All Customers {

    Customer {
        externalId='Optional[12345]'
        masterExternalId='Optional[12345]'
        companyNumber='Optional[470813-8895]'
        internalId='Optional[fake internalId]'
        name='Optional[Acme Inc.]'
        customerType=COMPANY
        preferredStore='Optional.empty'
        bonusPoints='Optional.empty'
        address='123 main st, SE-123 45 Helsingborg'
        shoppingLists=[[lipstick, blusher]]
    }

}
All Shopping Lists
[[lipstick, blusher]]
